packages/blockchain/src/consensus/clique.ts

```diff
@@ -0,0 +1,557 @@
+      this._cliqueLatestBlockSigners.push(signer)
+
+      // trim length to `this.cliqueSignerLimit()`
+      const length = this._cliqueLatestBlockSigners.length
+      const limit = this.cliqueSignerLimit()
+      if (length > limit) {
+        this._cliqueLatestBlockSigners = this._cliqueLatestBlockSigners.slice(
+          length - limit,
+          length
+        )
+      }
+    }
+
+    // save to db
+    const formatted = this._cliqueLatestBlockSigners.map((b) => [
+      bigIntToBuffer(b[0]),
+      b[1].toBuffer(),
+    ])
+    await this.blockchain.db.put('CliqueBlockSignersSnapshot', rlp.encode(formatted), DB_OPTS)
+  }
+
+  /**
+   * Fetches clique signers.
+   * @hidden
+   */
+  private async getCliqueLatestSignerStates(): Promise<CliqueLatestSignerStates> {
+    try {
+      const signerStates = await this.blockchain.db.get('CliqueSigners', DB_OPTS)
+      const states = (<any>rlp.decode(signerStates)) as [Buffer, Buffer[]]
+      return states.map((state) => {
+        const blockNum = bufferToBigInt(state[0] as Buffer)
+        const addrs = (<any>state[1]).map((buf: Buffer) => new Address(buf))
+        return [blockNum, addrs]
+      }) as CliqueLatestSignerStates
+    } catch (error: any) {
+      if (error.type === 'NotFoundError') {
+        return []
+      }
+      throw error
+    }
+  }
+
+  /**
+   * Fetches clique votes.
+   * @hidden
+   */
+  private async getCliqueLatestVotes(): Promise<CliqueLatestVotes> {
+    try {
+      const signerVotes = await this.blockchain.db.get('CliqueVotes', DB_OPTS)
+      const votes = (<any>rlp.decode(signerVotes)) as [Buffer, [Buffer, Buffer, Buffer]]
+      return votes.map((vote) => {
+        const blockNum = bufferToBigInt(vote[0] as Buffer)
+        const signer = new Address((vote[1] as any)[0])
+        const beneficiary = new Address((vote[1] as any)[1])
+        const nonce = (vote[1] as any)[2]
+        return [blockNum, [signer, beneficiary, nonce]]
+      }) as CliqueLatestVotes
+    } catch (error: any) {
+      if (error.type === 'NotFoundError') {
+        return []
+      }
+      throw error
+    }
+  }
+
+  /**
+   * Fetches snapshot of clique signers.
+   * @hidden
+   */
+  private async getCliqueLatestBlockSigners(): Promise<CliqueLatestBlockSigners> {
+    try {
+      const blockSigners = await this.blockchain.db.get('CliqueBlockSignersSnapshot', DB_OPTS)
+      const signers = (<any>rlp.decode(blockSigners)) as [Buffer, Buffer][]
+      return signers.map((s) => {
+        const blockNum = bufferToBigInt(s[0] as Buffer)
+        const signer = new Address(s[1] as any)
+        return [blockNum, signer]
+      }) as CliqueLatestBlockSigners
+    } catch (error: any) {
+      if (error.type === 'NotFoundError') {
+        return []
+      }
+      throw error
+    }
+  }
+
+  /**
+   * Gets a header by number. Header must be in the canonical chain
+   * @hidden
+   */
+  private async _getCanonicalHeader(number: bigint) {
```

Comment:
why redefine this method, can just use from the blockchain class?

Reply:
Good catch - this was also an early change. I changed it in the blockchain class to a public method `getCanonicalHeader` and removed the duplicate methods in `consensus/clique`